### ISSUE TITLE
net: Interfaces, InterfaceByIndex, InterfaceByName regression in 1.5 on windows

### ISSUE THREAD

**Issue by AudriusButkevicius** [golang/go#12301]
```
package main

import(
	"fmt"
	"net"
)

func main(){
	ifs, err := net.Interfaces()
	if err != nil {
		fmt.Println(err)
	}
	for _, inf := range ifs {
		x, _ := inf.Addrs()
		fmt.Printf("%#v %s", inf, x)
	}
}
```

In 1.4.1 returns:
```
net.Interface{Index:5, MTU:1500, Name:"{E36EC396-D578-48C0-82D5-F55217877570}", HardwareAddr:net.HardwareAddr{0x94, 0xde, 0x80, 0xaf, 0x4b, 0x62}, Flags:0x0} [192.168.1.70]
```

In 1.5 returns:
```
net.Interface{Index:1, MTU:4294967295, Name:"Loopback Pseudo-Interface 1", HardwareAddr:net.HardwareAddr{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}, Flags:0x15} [::1/28 127.0.0.1/16]
```

Any code relying on net.Interfaces() breaks (for example making multicasts to all interfaces, since Windows by default multicasts to some single interface)

What version of Go are you using (go version)?
go version go1.5 windows/amd64

What operating system and processor architecture are you using?
Windows 8.1 x64

What did you do?
Called net.Interfaces()

What did you expect to see?
My one external interface

What did you see instead?
My one loopback interface

ipconfig /all
```
c:\Go\src\net>ipconfig /all

Windows IP Configuration

   Host Name . . . . . . . . . . . . : Audrius
   Primary Dns Suffix  . . . . . . . :
   Node Type . . . . . . . . . . . . : Hybrid
   IP Routing Enabled. . . . . . . . : No
   WINS Proxy Enabled. . . . . . . . : No
   DNS Suffix Search List. . . . . . : lan

Ethernet adapter Ethernet 5:

   Connection-specific DNS Suffix  . : lan
   Description . . . . . . . . . . . : Hyper-V Virtual Ethernet Adapter #2
   Physical Address. . . . . . . . . : 94-DE-80-AF-4B-62
   DHCP Enabled. . . . . . . . . . . : Yes
   Autoconfiguration Enabled . . . . : Yes
   Link-local IPv6 Address . . . . . : fe80::58d9:707e:1173:32c%5(Preferred)
   IPv4 Address. . . . . . . . . . . : 192.168.1.70(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.255.0
   Lease Obtained. . . . . . . . . . : 23 August 2015 13:46:11
   Lease Expires . . . . . . . . . . : 25 August 2015 13:46:22
   Default Gateway . . . . . . . . . : 192.168.1.254
   DHCP Server . . . . . . . . . . . : 192.168.1.254
   DHCPv6 IAID . . . . . . . . . . . : 1150606976
   DHCPv6 Client DUID. . . . . . . . : 00-01-00-01-1D-3C-8E-0C-94-DE-80-AF-4B-62
   DNS Servers . . . . . . . . . . . : fe80::9e97:26ff:fe78:cec2%5
                                       8.8.8.8
                                       4.4.8.8
   NetBIOS over Tcpip. . . . . . . . : Enabled

Tunnel adapter isatap.lan:

   Media State . . . . . . . . . . . : Media disconnected
   Connection-specific DNS Suffix  . : lan
   Description . . . . . . . . . . . : Microsoft ISATAP Adapter #4
   Physical Address. . . . . . . . . : 00-00-00-00-00-00-00-E0
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes

```

netsh interface show interface
```

Admin State    State          Type             Interface Name
-------------------------------------------------------------------------
Enabled        Connected      Dedicated        Ethernet 3
Enabled        Connected      Dedicated        Ethernet 5
```

Test output:
```
c:\Go\src\net>go test -v -run=Interface
=== RUN   TestInterfaces
--- PASS: TestInterfaces (0.00s)
        interface_test.go:79: "Loopback Pseudo-Interface 1": flags "up|loopback|multicast", ifindex 1, mtu 4294967295
        interface_test.go:80: hardware address "00:00:00:00:00:00:00:00"
        interface_test.go:202: interface address "::1/28"
        interface_test.go:202: interface address "127.0.0.1/16"
        interface_test.go:223: joined group address "ff02::c"
        interface_test.go:223: joined group address "239.255.255.250"
=== RUN   TestInterfaceAddrs
--- PASS: TestInterfaceAddrs (0.00s)
        interface_test.go:202: interface address "fe80::58d9:707e:1173:32c/28"
        interface_test.go:202: interface address "192.168.1.70/16"
        interface_test.go:202: interface address "::1/28"
        interface_test.go:202: interface address "127.0.0.1/16"
        interface_test.go:202: interface address "fe80::5efe:c0a8:146/28"
PASS
Socket statistical information:
(inet4, datagram, udp): opened=4, connected=0, listened=0, accepted=0, closed=4, openfailed=0, connectfailed=0, listenfailed=0, acceptfailed=0, closefailed=0

ok      net     0.124s
```

**Comment by AudriusButkevicius** [comment]
Strangely, `net.InterfaceAddrs()` returns the correct addresses, though `net.InterfaceByIndex()` not 'net.InterfaceByName()' does not.

**Comment by AudriusButkevicius** [comment]
Seems that it has been broken by https://github.com/golang/go/commit/ea22a08fd4f8ebb296baf4ff59861d8885ac3435 by @mattn.

The problem seems to lie in `getInterfaceInfos()` which only returns info for the loopback adapter.

**Comment by mikioh** [comment]
@AudriusButkevicius,

Please take a look at https://github.com/golang/go/blob/master/CONTRIBUTING.md and let us know your environment information. Also can you show us the output of "go test -v -run=Interface" and "ipconfig? ifconfig? netsh interfaces show?" on the node? (Please anonymize IDs in the output if necessary)

**Comment by AudriusButkevicius** [comment]
I edited my original post.

**Comment by mikioh** [comment]
Thanks. Just skimmed interface_windows.go and looks like the code is broken. At least about the following:

- It doesn't list up all equipped interfaces because of https://github.com/golang/go/blob/master/src/net/interface_windows.go#L110
- It makes wrong address prefixes because of https://github.com/golang/go/blob/master/src/net/interface_windows.go#L167. It should take prefixes instead of addresses

**Comment by AudriusButkevicius** [comment]
Re 1, it does, because of `for ; paddr != nil; paddr = paddr.Next {`

In my case:
https://github.com/golang/go/blob/master/src/net/interface_windows.go#L93
chain contains 3 interfaces, but https://github.com/golang/go/blob/master/src/net/interface_windows.go#L98 contains only one info, hence interfaces get skipped at
https://github.com/golang/go/blob/master/src/net/interface_windows.go#L112

**Comment by alexbrainman** [comment]
@mikioh feel free to change. I know nothing about these. Happy to test / review if you need my help.

Alex

**Comment by mikioh** [comment]
@AudriusButkevicius,

Feel free to send a fix if you have a spare time. Unfortunately I'm not a windows user, have no windows box.

**Comment by mikioh** [comment]
Ah... @AudriusButkevicius, after you.

**Comment by AudriusButkevicius** [comment]
Well the win32 api is not clear to me, why it would return only info for a single interface, hence I am not sure what the right answer here is, hence pinging @mattn in hopes for comments.

**Comment by mattn** [comment]
I can reproduce. I'll look into it in later. But I don't have the old environment windowns XP 32bit that I made the change.no longer. Currently, I'm on windows 64bit.

**Comment by AudriusButkevicius** [comment]
Well for me I can reproduce on win8.1 x64, so shouldn't be a problem.

**Comment by mattn** [comment]
Hmm, I'm trying this on another 64bit windows. But test are passed. I'm confusing.

**Comment by rsc** [comment]
Why did this change from 1.4?

**Comment by AudriusButkevicius** [comment]
So tests are passing too

**Comment by alexbrainman** [comment]
> Why did this change from 1.4?

http://golang.org/cl/3024

I don't understand most details. We just tried to make tests pass.

Alex

**Comment by mikioh** [comment]
@mattn,

If you have no spare time for now, I'll take a look a bit this weekend.

> Why did this change from 1.4?

Old windows APIs used in the previous versions don't support IPv6 entirely.

**Comment by mattn** [comment]
> Old windows APIs used in the previous versions don't support IPv6 entirely.

Yes, And some APIs doesn't support wide character set.

> If you have no spare time for now, I'll take a look a bit this weekend.

Do it please. But I'll look if it possible.

**Comment by 0086** [comment]
The function getInterfaceInfos() in "src/net/interface_windows.go" return a incorrect slice, 
``` go
return iia[:iilen-1], nil
```
This line code should be 
``` go
return iia[:iilen], nil
```

**Comment by AudriusButkevicius** [comment]
I'll test this later tonight.

**Comment by AudriusButkevicius** [comment]
:+1:  this fixes it.
So sad that it has missed 1.5.1

**Comment by calmh** [comment]
:heart: Thank you.

**Comment by mattn** [comment]
@AudriusButkevicius sorry and sad :(

**Comment by AudriusButkevicius** [comment]
I see this hasn't been closed yet, though the changeset looked complete. Any issues along the way? 

I really hope this makes it for 1.5.2.

**Comment by alexbrainman** [comment]
>  see this hasn't been closed yet, though the changeset looked complete. Any issues along the way?

CL https://golang.org/cl/14747 needs a test that won't let us break that code again. Also @mikioh proposed an alternative solution https://go-review.googlesource.com/#/c/14747/

Alex

**Comment by AudriusButkevicius** [comment]
Any way I can help you guys (@mikioh, @mattn) to get this fixed? I don't want to miss tye 1.5.2 train.

**Comment by rsc** [comment]
OK for Go 1.5.2.
